File: modules/abquad.c
#include <stdlib.h>
#include <math.h>
#include "soundpipe.h"

int sp_abquad_create(sp_abquad **p)
{
		*p = malloc(sizeof(sp_abquad));
		return SP_OK;
}

int sp_abquad_destroy(sp_abquad **p)
{
		free(*p);
		return SP_OK;
}

int sp_abquad_init(sp_data* sp, sp_abquad* p)
{
		p->b0 = 0;
		p->b1 = 0;
		p->b2 = 0;
		p->a1 = 0;
		p->a2 = 0;
		p->z1 = 0;
		p->z2 = 0;
		return SP_OK;
}

int sp_abquad_compute(sp_data* sp, sp_abquad* p, SPFLOAT *input, SPFLOAT* output)
{
		SPFLOAT in = *input;
	  *output = p->z1 + in * p->b0;
	  p->z1 = p->z2 + in * p->b1 - *output*p->a1;
	  p->z2 = in * p->b2 - *output*p->a2;
		return SP_OK;
}

int sp_abquad_set(sp_data* sp, sp_abquad* p, const SPFLOAT coefs[5])
{
		p->b0 = coefs[0];
		p->b1 = coefs[1];
		p->b2 = coefs[2];
		p->a1 = coefs[3];
		p->a2 = coefs[4];
		return SP_OK;
}

File: modules/data/abquad.lua
sptbl["abquad"] = {

    files = {
        module = "abquad.c",
        header = "abquad.h",
        --example = "ex_abquad.c",
    },

    func = {
        create = "sp_abquad_create",
        destroy = "sp_abquad_destroy",
        init = "sp_abquad_init",
        compute = "sp_abquad_compute",
        other = {
            sp_abquad_set = {
                description = "set coefs at once.",
                args = {
				            {
  		  		            name = "b0",
    		  		          type = "SPFLOAT",
      		  		        description = "biquad coefficient.",
  		        		      default = 0
  		          		},
    		        		{
       		         		name = "b1",
          		      		type = "SPFLOAT",
              		  		description = "biquad coefficient.",
                				default = 0
		            		},
    		        		{
        		        		name = "b2",
            		    		type = "SPFLOAT",
                				description = "biquad coefficient.",
                				default = 0
		            		},
    		        		{
        		        		name = "a1",
            		    		type = "SPFLOAT",
                				description = "biquad coefficient.",
                				default = 0
		            		},
    		        		{
        		        		name = "a2",
            		    		type = "SPFLOAT",
                				description = "biquad coefficient.",
                				default = 0
		            		}
                }
            }
        }

    },

    params = {
        optional = {
            {
                name = "b0",
                type = "SPFLOAT",
                description = "biquad coefficient.",
                default = 0
            },
            {
                name = "b1",
                type = "SPFLOAT",
                description = "biquad coefficient.",
                default = 0
            },
            {
                name = "b2",
                type = "SPFLOAT",
                description = "biquad coefficient.",
                default = 0
            },
            {
                name = "a1",
                type = "SPFLOAT",
                description = "biquad coefficient.",
                default = 0
            },
            {
                name = "a2",
                type = "SPFLOAT",
                description = "biquad coefficient.",
                default = 0
            }
            {
                name = "z1",
                type = "SPFLOAT",
                description = "biquad coefficient.",
                default = 0
            },
            {
                name = "z2",
                type = "SPFLOAT",
                description = "biquad coefficient.",
                default = 0
            },
        }
    },

    modtype = "module",

    description = [[A sweepable biquadratic general purpose filter

    ]],

    ninputs = 1,
    noutputs = 1,

    inputs = {
        {
            name = "input",
            description = "Signal input."
        },
    },

    outputs = {
        {
            name = "out",
            description = "Signal output."
        },
    }

}
